update error code docs to reflect recent changes (#1625)
rust-lang/rustc-dev-guide: src/diagnostics.md

it's a good idea to add an explanation.[^estebank]
As always, if you are not sure, just ask your reviewer!

> If you do decide to add a error code with a new error (quite likely), please

Reviewer: ```suggestion
If you decide to add a new error with an associated error code, please
```
